update correlation qc

ErinWeisbart opened the pull request:
minor expansion to Correlation QC visualization:
- makes spreadsheet a bit easier to view at a glance
- adds by-well plate view of failed correlations

added in code so that site with NA values don't plot on plate overviews of {compartment}_thresholds and confluence.

gwaygenomics commented on 0.preprocess-sites/4.image-and-segmentation-qc.py:
```diff
@@ -490,7 +494,7 @@
 # Create list of questionable channel correlations (alignments)
 correlation_col_prefix = "Correlation_Correlation_"
-corr_base_cols = image_meta_col_list.copy() + ["site_location"]
+corr_base_cols = image_meta_col_list.copy() + ["site"]
```
I don't think we want this change right? Our plots should include the number location, not the full string, right?  discussed https://github.com/broadinstitute/pooled-cell-painting-profiling-recipe/issues/45#issuecomment-670206381

ErinWeisbart replied:
Drats. Sorry about that. I had switched to an older version so I could quickly generate qc metrics and then forgot to switch back. You are correct.

gwaygenomics commented on 0.preprocess-sites/4.image-and-segmentation-qc.py:
```diff
@@ -504,14 +508,48 @@
+        pd.concat(image_corr_list).drop_duplicates(subset="site").reset_index()
+    )
+
+    image_corr_df["min_corr_value"] = (
```
is this creating a new variable? For consistency, lets use the notation:

```python
image_corr_df = image_corr_df.assign(min_corr_value = ...)
```

Also, the new column is "min_corr_value" but we're taking the max? Am i reading this right?

ErinWeisbart replied:
That's a mistake. Fixed max => min.

gwaygenomics commented on 0.preprocess-sites/4.image-and-segmentation-qc.py:
```diff
@@ -504,14 +508,48 @@
     for col in corr_qc_cols:
         image_corr_df.loc[(image_corr_df[col] >= correlation_threshold), col] = "pass"
 
+    for col in corr_qc_cols:
```
I am confused about what this is doing. Also, is there a reason to loop over `corr_qc_cols` twice? Or can these operations be combined in the same loop?

ErinWeisbart replied:
There is no need to loop twice. I've combined them now. 
I'm doing some perhaps un-intuitive manipulation so that the resulting .csv is easy for a human to extract information at a glance. The goal:
We have correlation scores for each site for each pairwise image combo. 
It looks for anything where the correlation is good enough (>= correlation_threshold) and changes the value to "pass" since if the sites are aligned, we don't really care at a glance how good the good is.
Then it looks for any column where there are no alignment problems (if all "pass" then len(image_corr_df[col].unique()) == 1) and drops those columns.
So the resulting spreadsheet only contains sites with alignment problems and only scores for image pairs that have problems.

gwaygenomics replied:
if all correlations fail, will `len(image_corr_df[col].unique()) == 1` be `True`?

ErinWeisbart replied:
no. there is effectively no way that the correlations would all fail with the same score.

gwaygenomics commented on 0.preprocess-sites/4.image-and-segmentation-qc.py:
```diff
@@ -504,14 +508,48 @@
+
+# By-well visualization of sites with poor correlations
+if len(corr_qc_cols) > 0:
+    image_corr_df = image_corr_df.merge(loc_df, on="site")
```
check "site" notation - not sure about `loc_df` but I do remember fixing in d98ded387fe376ea310ae23847ea40758de2c388

ErinWeisbart replied:
yep, sorry, another casualty of my use of an older version.

gwaygenomics commented on 0.preprocess-sites/4.image-and-segmentation-qc.py:
```diff
@@ -504,14 +508,48 @@
+            axis_title=gg.element_blank(),
+            strip_background=gg.element_rect(colour="black", fill="#fdfff4"),
+        )
+        + gg.labs(fill="Minimum correlation in image set")
```
Can we be slightly more specific in the plot title and/or axis labels? Is this visualizing the dapi/hoechst correlation between ISS and cell painting?

ErinWeisbart replied:
It's a plate overview so I'm not sure what axis labels would add?
This is looking at the scores for all pairwise image correlations (not just DAPI between BC and CP). It only plots sites that have image correlations that fail to pass the QC threshold and heatmaps the worst pairwise correlation of the image set. It's not a generally super informative or necessary plot, but it would help us visually catch the circumstance where 1) one region of the well is aligning particularly badly or 2) images after a certain point are all aligning badly.

ErinWeisbart replied:
That being said, I'm still not sure what a more informative title would be, and open to suggestions!

gwaygenomics replied:
Ah got it. I think I was being thrown off by the title "Failed Correlations". Correlations between what? How about "Failed Image Channel Correlations\n{plate}"? I will add as a suggestion, but feel free to further modify!

gwaygenomics commented on 0.preprocess-sites/4.image-and-segmentation-qc.py:
```diff
@@ -504,14 +508,48 @@
+        + gg.scale_fill_cmap(name="magma")
+    )
+
+    os.makedirs(figures_output, exist_ok=True)
```
I am pretty sure we define this line earlier, can probably delete

ErinWeisbart commented:
@gwaygenomics thanks for the review!
Sorry again about wasting your time on the `site` naming mistakes. Yes, they were from using an older version.
I think I've addressed all your suggestions, other than naming the new plot. Perhaps with my more thorough description of the goal of the plot you can help suggest a better name for it?

gwaygenomics commented on 0.preprocess-sites/4.image-and-segmentation-qc.py:
```diff
@@ -504,14 +509,47 @@
 
     image_corr_df = (
-        pd.concat(image_corr_list).drop_duplicates(subset="site_location").reset_index()
+        pd.concat(image_corr_list).drop_duplicates(subset="site").reset_index()
```
```suggestion
        pd.concat(image_corr_list).drop_duplicates(subset="site_location").reset_index()
```

Is this right?

ErinWeisbart replied:
No, I think it's correct with `site` not `site_location`.

gwaygenomics commented on 0.preprocess-sites/4.image-and-segmentation-qc.py:
```diff
@@ -504,14 +509,47 @@
+    corr_by_well_gg = (
+        gg.ggplot(image_corr_df, gg.aes(x="x_loc", y="y_loc"))
+        + gg.geom_point(gg.aes(fill="min_corr_value"), size=10)
+        + gg.geom_text(gg.aes(label="site"), color="lightgrey")
```
checkin that the text should indeed be "site" (include plate name) or "site_location" (just the number)

gwaygenomics replied:
sorry, should say "I am checking in to confirm that..." doh

ErinWeisbart replied:
All these `site` names have me really muddled. Looking forward to having v.1 locked down!
All of the plot labels should be just the number, so that's `site_location`

gwaygenomics replied:
me too! Thanks for bearing with these frustrating updates - we'll have a lock soon

gwaygenomics commented on 0.preprocess-sites/4.image-and-segmentation-qc.py:
```diff
@@ -504,14 +509,47 @@
+        + gg.facet_wrap("~Metadata_Well")
+        + gg.coord_fixed()
+        + gg.theme_bw()
+        + gg.ggtitle(f"Failed Correlations \n {plate}")
```
```suggestion
        + gg.ggtitle("Failed Image Channel Correlations\n{plate}")
```

gwaygenomics replied:
woah, this is my bad - we need to add back the `f`
